Libraries/libutil/Sources/DefaultFilesystem.cpp

```diff
@@ -194,53 +210,197 @@ readSymbolicLink(std::string const &path) const
+{
+    struct stat st;
+    if (::stat(path.c_str(), &st) < 0) {
+        return false;
     } else {
-        return realPath;
+        return S_ISDIR(st.st_mode);
+    }
+}
+
+bool DefaultFilesystem::
+createDirectory(std::string const &path, bool recursive)
+{
+    /* Get current mode. */
+    mode_t mask = ::umask(0);
+    ::umask(mask);
+
+    /* Mode is most allowed by mask. */
+    mode_t mode = (S_IRWXU | S_IRWXG | S_IRWXO) & ~mask;
+
+    if (recursive) {
+        std::string current = path;
+        std::stack<std::string> create;
+
+        /* Build up list of directories to create. */
+        while (!this->isDirectory(path)) {
+            create.push(current);
+            current = FSUtil::GetDirectoryName(current);
+        }
+
+        /* Create intermediate directories. */
+        while (!create.empty()) {
+            std::string const &directory = create.top();
+
+            if (::mkdir(directory.c_str(), mode) != 0) {
+                return false;
+            }
+
+            create.pop();
+        }
+    } else {
+        if (::mkdir(path.c_str(), mode) != 0) {
+            return false;
+        }
     }
+
+    return true;
 }
 
 bool DefaultFilesystem::
-enumerateDirectory(
-    std::string const &path,
-    std::function<void(std::string const &)> const &cb) const
+readDirectory(std::string const &path, bool recursive, std::function<void(std::string const &)> const &cb) const
 {
-    DIR *dp = opendir(path.c_str());
+    DIR *dp = ::opendir(path.c_str());
     if (dp == NULL) {
         return false;
     }
 
-    while (struct dirent *entry = readdir(dp)) {
+    while (struct dirent *entry = ::readdir(dp)) {
         std::string name = entry->d_name;
         if (name != "." && name != "..") {
             cb(name);
         }
     }
 
-    closedir(dp);
+    ::closedir(dp);
+    return true;
+}
+
+bool DefaultFilesystem::
+copyDirectory(std::string const &from, std::string const &to, bool recursive)
+{
+#if defined(__APPLE__) || defined(__FreeBSD__)
+    if (!this->isDirectory(from)) {
+        return false;
+    }
+
+    if (this->isDirectory(to)) {
+        if (!this->removeDirectory(to, recursive)) {
+            return false;
+        }
+    }
+
+    copyfile_state_t state = ::copyfile_state_alloc();
+    copyfile_flags_t flags = COPYFILE_ALL | COPYFILE_NOFOLLOW | (recursive ? COPYFILE_RECURSIVE : 0);
+    if (::copyfile(from.c_str(), to.c_str(), state, flags)) {
+        return false;
+    }
+    ::copyfile_state_free(state);
+
+    return true;
+#else
+    return Filesystem::copyDirectory(from, to, recursive);
+#endif
+}
+
+bool DefaultFilesystem::
+removeDirectory(std::string const &path, bool recursive)
+{
+    if (recursive) {
+        bool success = true;
+
+        this->readDirectory(path, recursive, [this, &path, &success](std::string const &name) {
+            std::string full = path + "/" + name;
+            if (this->isDirectory(full)) {
+                if (!this->removeDirectory(full, false)) {
```

Comment:
The outer `readDirectory()` is recursing, so no need to recurse again on the inside, it'll handle it.